tf_trees/tel.py

```diff
@@ -0,0 +1,190 @@
+                 parallelize_over_samples=False,
+                 split_initializer=RandomUniform(-0.01, 0.01),
+                 leaf_initializer=RandomUniform(-0.01, 0.01),
+                 split_regularizer=None,
+                 leaf_regularizer=None,
+                 **kwargs):
+      """Initializes neural trees layer."""
+      self.trees_num = trees_num
+      self.depth = depth
+      self.output_logits_dim = output_logits_dim
+      self.smooth_step_param = smooth_step_param
+      self.parallelize_over_samples = parallelize_over_samples
+      self.sum_outputs = sum_outputs
+      self.split_initializer = keras.initializers.get(split_initializer)
+      self.leaf_initializer = keras.initializers.get(leaf_initializer)
+      self.split_regularizer = keras.regularizers.get(split_regularizer)
+      self.leaf_regularizer = keras.regularizers.get(leaf_regularizer)
+      super(TEL, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+      """Creates a keras layer."""
+      num_leaves = 2**self.depth
+      num_internal_nodes = num_leaves - 1
+      input_dim = input_shape[1]
+      # A list of node weights. Each element corresponds to one tree.
+      self.node_weights = []
+      # A list of leaf weights. Each element corresponds to one tree.
+      self.leaf_weights = []
+      # Create node/leaf weights for every tree.
+      for _ in range(self.trees_num):
+        self.node_weights.append(
+            self.add_weight(
+                name='node_weights',
+                shape=[input_dim, num_internal_nodes],
+                initializer=self.split_initializer,
+                trainable=True,
+                regularizer=self.split_regularizer))
+
+        self.leaf_weights.append(
+            self.add_weight(
+                name='leaf_weights',
+                shape=[self.output_logits_dim, num_leaves],
+                initializer=self.leaf_initializer,
+                trainable=True,
+                regularizer=self.leaf_regularizer))
+
+        self.smooth_step_param_var = self.add_weight(
+                name='smooth_step_param_var',
+                shape=[],
+                initializer =
+                tf.constant_initializer(float(self.smooth_step_param)),
+                trainable=False)
+      super(TEL, self).build(input_shape)
+
+    def call(self, inputs):
+      """Predict op."""
+      # A list of tree outputs. Each element corresponds to one tree.
+      tree_logits = []
+      for tree_index in range(self.trees_num):
+        tree_out, _ = tf_trees_module.nt_compute_output_op(inputs,
+                                 self.node_weights[tree_index],
+                                 self.leaf_weights[tree_index],
+                                 self.smooth_step_param_var,
+                                 self.output_logits_dim, self.depth,
+                                 self.parallelize_over_samples)
+        tree_logits.append(tree_out)
+      if self.trees_num == 1:
```

Comment:
may be don't do it. imagine they dont want to sum and they tune between different number of trees and the code expects concated logits at this point. Just return tree_logits if do not accumulate